apps/src/music/views/PatternAiPanel.tsx

```diff
@@ -0,0 +1,380 @@
+  };
+
+  const handleFolderChange = (event: ChangeEvent<HTMLSelectElement>) => {
+    currentValue.kit = event.target.value;
+    onChange(currentValue);
+  };
+
+  const getCellClasses = (sound: SoundData, tick: number) => {
+    const isSeed = tick < 9;
+    const isHighlighted = (tick - 1) % 4 === 0;
+    const isActive = hasEvent(sound, tick);
+    const isPlaying = isActive && tick === currentPreviewTick;
+
+    return classNames(
+      styles.cell,
+      isSeed && isActive
+        ? styles.cellSeedActive
+        : isSeed && isHighlighted
+        ? styles.cellSeedHighlighted
+        : isSeed
+        ? styles.cellSeedDefault
+        : userCompletedTask !== 'generated'
+        ? styles.cellGeneratedInvisible
+        : !isSeed && isActive
+        ? styles.cellGeneratedActive
+        : !isSeed && isHighlighted
+        ? styles.cellGeneratedHighlighted
+        : styles.cellGeneratedDefault,
+      isPlaying && styles.cellPlaying
+    );
+  };
+
+  const onClear = useCallback(() => {
+    currentValue.events = [];
+    onChange(currentValue);
+  }, [onChange, currentValue]);
+
+  useEffect(() => {
+    if (!isInstrumentLoaded(currentValue.kit)) {
+      setIsLoading(true);
+      if (isInstrumentLoading(currentValue.kit)) {
+        // If the instrument is already loading, register a callback and wait for it to finish.
+        registerInstrumentLoadCallback(kit => {
+          if (kit === currentValue.kit) {
+            setIsLoading(false);
+          }
+        });
+      } else {
+        // Otherwise, initiate the load.
+        setupSampler(currentValue.kit, () => setIsLoading(false));
+      }
+    }
+  }, [
+    setupSampler,
+    isInstrumentLoading,
+    isInstrumentLoaded,
+    currentValue.kit,
+    setIsLoading,
+    registerInstrumentLoadCallback,
+  ]);
+
+  // Tracks the tasks completed by the user.
+  useEffect(() => {
+    if (currentValue.events.some(event => event.tick >= 9)) {
+      setUserCompletedTask('generated');
+    } else if (currentValue.events.length >= 4) {
+      if (userCompletedTask === 'none') {
+        setUserCompletedTask('drawnDrums');
+      }
+    }
+  }, [currentValue.events, userCompletedTask]);
+
+  const startPreview = useCallback(
+    (value: PatternEventValue) => {
+      previewPattern(
+        value,
+        (tick: number) => {
+          setCurrentPreviewTick(tick);
+        },
+        () => {
+          setCurrentPreviewTick(0);
+        }
+      );
+    },
+    [previewPattern, setCurrentPreviewTick]
+  );
+
+  const stopPreview = useCallback(() => {
+    cancelPreviews();
+    setCurrentPreviewTick(0);
+  }, [cancelPreviews]);
+
+  const playPreview = useCallback(() => {
+    startPreview(currentValue);
+  }, [startPreview, currentValue]);
+
+  const handleAiClick = useCallback(async () => {
+    stopPreview();
+    const seedEvents = currentValue.events.filter(event => event.tick <= 8);
+    generatePattern(seedEvents, 8, 32 - 8, aiTemperature, newEvents => {
```

Comment:
mind creating some constants and/or add comments for what these numbers mean?

Reply:
Good call; done.